staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/typeconverter.go

```diff
@@ -35,25 +37,59 @@ type TypeConverter interface {
 	TypedToObject(*typed.TypedValue) (runtime.Object, error)
 }
 
+type typeResolver interface {
+	Type(schema.GroupVersionKind) *typed.ParseableType
+}
+
+type mapTypeResolver map[schema.GroupVersionKind]*typed.ParseableType
+
+func (m mapTypeResolver) Type(k schema.GroupVersionKind) *typed.ParseableType {
+	return m[k]
+}
+
 type typeConverter struct {
-	parser *managedfields.GvkParser
+	parser typeResolver
 }
 
 var _ TypeConverter = &typeConverter{}
 
-// NewTypeConverter builds a TypeConverter from a proto.Models. This
+// NewTypeConverter builds a TypeConverter from a spec.Swagger. This
 // will automatically find the proper version of the object, and the
 // corresponding schema information.
 func NewTypeConverter(openapiSpec *spec.Swagger, preserveUnknownFields bool) (TypeConverter, error) {
-	models, err := utilopenapi.ToProtoModels(openapiSpec)
-	if err != nil {
-		return nil, err
+	if openapiSpec == nil {
+		return nil, errors.New("openapi spec must not be nil")
 	}
-	parser, err := managedfields.NewGVKParser(models, preserveUnknownFields)
+
+	pointerDefs := map[string]*spec.Schema{}
+	for k, v := range openapiSpec.Definitions {
+		vCopy := v
+		pointerDefs[k] = &vCopy
+	}
+
+	typeSchema, err := schemaconv.ToSchemaFromOpenAPI(pointerDefs, preserveUnknownFields)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to convert models to schema: %v", err)
+	}
+
+	typeParser := typed.Parser{Schema: smdschema.Schema{Types: typeSchema.Types}}
+
+	tr := mapTypeResolver{}
+	for modelName, model := range pointerDefs {
+		gvkList := parseGroupVersionKind(model.Extensions)
+		if len(gvkList) == 0 {
+			continue
+		}
+
+		parsedType := typeParser.Type(modelName)
+		for _, gvk := range gvkList {
+			if len(gvk.Kind) > 0 {
+				tr[schema.GroupVersionKind(gvk)] = &parsedType
+			}
+		}
 	}
```

Comment:
maybe move this logic out of that function, just for isolation (i.e. you would be able to test that function individually).

Reply:
maybe we should restore the gvk parser then?

Reply:
moved it out and added test